tests/test_fgnhg.py

```diff
@@ -626,12 +643,126 @@ def test_route_fgnhg(self, dvs, testlog):
+            remove_entry(config_db, IF_TB, ip_pref_key)
+            remove_entry(config_db, IF_TB, vlan_name_key)
+            dvs.runcmd("config interface shutdown " + if_name_key)
+            dvs.servers[i].runcmd("ip link set down dev eth0") == 0
+            remove_entry(config_db, "FG_NHG_MEMBER", "10.0.0." + str(1 + i*2))
 
-        # Remove route
-        # remove prefix entry
-        asic_rt_key = get_asic_route_key(asic_db, fg_nhg_prefix)
-        ps._del(fg_nhg_prefix)
-        asic_db.wait_for_deleted_entry(ASIC_ROUTE_TB, asic_rt_key)
-        asic_db.wait_for_n_keys(ASIC_NHG_MEMB, 0)
+    def test_route_fgnhg_warm_reboot(self, dvs, testlog):
+        dvs.runcmd("swssloglevel -l INFO -c orchagent")
+        app_db = dvs.get_app_db()
+        asic_db = dvs.get_asic_db()
+        config_db = dvs.get_config_db()
+        state_db = dvs.get_state_db()
+        fvs_nul = {"NULL": "NULL"}
+        NUM_NHs = 6
+        fg_nhg_name = "fgnhg_v4"
+        fg_nhg_prefix = "2.2.2.0/24"
+        bucket_size = 60
+        ip_to_if_map = {}
 
-        remove_entry(config_db, "FG_NHG_PREFIX", fg_nhg_prefix)
+        fvs = {"bucket_size": str(bucket_size)}
+        create_entry(config_db, FG_NHG, fg_nhg_name, fvs)
+
+        fvs = {"FG_NHG": fg_nhg_name}
+        create_entry(config_db, FG_NHG_PREFIX, fg_nhg_prefix, fvs)
+
+        for i in range(0,NUM_NHs):
+            if_name_key = "Ethernet" + str(i*4)
+            vlan_name_key = "Vlan" + str((i+1)*4)
+            ip_pref_key = vlan_name_key + "|10.0.0." + str(i*2) + "/31"
+            fvs = {"vlanid": str((i+1)*4)}
+            create_entry(config_db, VLAN_TB, vlan_name_key, fvs)
+            fvs = {"tagging_mode": "untagged"}
+            create_entry(config_db, VLAN_MEMB_TB, vlan_name_key + "|" + if_name_key, fvs)
+            create_entry(config_db, VLAN_IF_TB, vlan_name_key, fvs_nul)
+            create_entry(config_db, VLAN_IF_TB, ip_pref_key, fvs_nul)
+            dvs.runcmd("config interface startup " + if_name_key)
+            dvs.servers[i].runcmd("ip link set down dev eth0") == 0
+            dvs.servers[i].runcmd("ip link set up dev eth0") == 0
+            bank = 0
+            if i >= NUM_NHs/2:
+                bank = 1
+            fvs = {"FG_NHG": fg_nhg_name, "bank": str(bank), "link": if_name_key}
+            create_entry(config_db, FG_NHG_MEMBER, "10.0.0." + str(1 + i*2), fvs)
+            ip_to_if_map["10.0.0." + str(1 + i*2)] = vlan_name_key
+
+        # Wait for the software to receive the entries
+        time.sleep(1)
+
+        ps = swsscommon.ProducerStateTable(app_db.db_connection, ROUTE_TB)
+        fvs = swsscommon.FieldValuePairs([("nexthop","10.0.0.7,10.0.0.9,10.0.0.11"),
+            ("ifname", "Vlan16,Vlan20,Vlan24")])
+        ps.set(fg_nhg_prefix, fvs)
+        # No ASIC_DB entry we can wait for since ARP is not resolved yet,
+        # We just use sleep so that the sw receives this entry
+        time.sleep(1)
```

Comment:
can you get rid of such sleep and wait_till?

Reply:
This sleep should be harmless since there isn't any state/asic db obj to check for after the sleep. This logic is the same on master today, and have not seen flakiness with test because of sleep. I added the sleep so that the sw can receive the entries and trigger a retry logic for the route. Even if I remove the sleep the outcome of this test is the same. The advantage of sleep is that it gives more test coverage here by allowing route retry logic to kick in, but even if retry does not occur, there should not be flakiness occurring due to it.

Reply:
what about the sleep above?